modules/runner/src/testFixtures/java/org/apache/ignite/internal/BaseIgniteRestartTest.java

```diff
@@ -139,15 +140,27 @@ void setUp(TestInfo testInfo) {
     public void afterEachTest() throws Exception {
         var closeables = new ArrayList<AutoCloseable>();
 
-        for (IgniteServer node : IGNITE_SERVERS) {
+        List<String> serverNames = IGNITE_SERVERS.stream()
+                .filter(Objects::nonNull)
+                .map(IgniteServer::name)
+                .collect(toList());
+
+        log.info("Shutting the cluster down [nodes={}]", serverNames);
+
+        // Stop all nodes in reverse order to ensure that the first started node (MS node) is stopped last.
+        for (int i = IGNITE_SERVERS.size() - 1; i >= 0; i--) {
+            IgniteServer node = IGNITE_SERVERS.get(i);
+
             if (node != null) {
                 closeables.add(node::shutdown);
             }
         }
 
         if (!partialNodes.isEmpty()) {
-            for (PartialNode partialNode : partialNodes) {
-                closeables.add(partialNode::stop);
+            for (int i = partialNodes.size() - 1; i >= 0; i--) {
```

Comment:
Provided solution is not robust. Generally (as a hack) we should stop CMG/MG last, while we may have CMG/MG hosted on both partial and full nodes, thus it's not valid to stop IGNITE_SERVERS first and partialNodes after that.